movies/3.api-doc.md

```diff
@@ -0,0 +1,89 @@
+# The API documentation
+
+## Story
+
+As a client developer, I want to find human-readable documenation of the API.
+As a client, I want to find a machine-readable description of the API capabilities.  
+
+## API behaviour
+
+```http
+GET https://hydra-movies.herokuapp.com/
+```
+
+Response headers (excerpt):
+
+```http
+Link: </doc>; rel="http://www.w3.org/ns/hydra/core#apiDocumentation"
+```
+
+```http
+GET https://hydra-movies.herokuapp.com/doc
+```
+
+Response body (excerpt):
+
+```json
+{
+  "@context": "/context.jsonld",
+  "@id": "/doc",
+  "@type": "ApiDocumentation",
+  "title": "Hydra Movies API",
+  "description": "This API shows how a basic CRUD (Create, Read, Update, Delete) API can be designed using Hydra Core Vocabulary.",
+  "entrypoint": "/",
+  "supportedClass": [
+    {
+      "@id": "schema:EntryPoint",
+      "@type": "Class",
+      "title": "API entrypoint",
+      "description": "Every capability of the API can be reached from here by following links."
+    }
+  ]
+}
+```
+
+### Explanation
+
+```http
+Link: </doc>; rel="http://www.w3.org/ns/hydra/core#apiDocumentation"
+```
+
+```http
+GET https://hydra-movies.herokuapp.com/doc
+```
+
+As described in ["API entrypoint"](./1.entry-point.md), every API response has to contain a Link to the documentation. This allows a client to dereference and process the documentation, even if it does not start from the entrypoint (e.g. because of a "deep link" or a bookmark from a former visit).
+
+```
+"@context": "/context.jsonld",
+"@id": "/doc",
+"@type": "ApiDocumentation"
+```
+
+The API documenation is just another web rescoure, with of course an `@id` (i.e. URI) and the type `http://www.w3.org/ns/hydra/core#ApiDocumentation`, which is abbreviated to `ApiDocumentation` here thanks to the JSON-LD context.
+
+```
+"title": "Hydra Movies API",
+"description": "This API shows how a basic CRUD (Create, Read, Update, Delete) API can be designed using Hydra Core Vocabulary.",
+```
+
+`title` and `description` are hydra terms that allow to describe the API in a human-readable matter on a high level. The values provided here, could for example appear in generic clients, API catalogs or API search engine results. They allow humans to get a first impression of what the API is good for.
+
+```
+"entrypoint": "/",
+```
+
+Since a client may get to the API by any URL, not only the entry point, it is important to make clear what the entry point of the API actually is. The entry point may also differ by the state of a client: An authenticated client could for example be sent to a user profile, while anonymous clients get a generic entry point. 
```

Comment:
> The entry point may also differ by the state of a client

This one is interesting. I remember we had a discussion on possibility of having more than one entry point. I stated that for some circumstances (i.e. one API being merged with another) it might be the case. Here it comes to the next level. And I'm not sure whether it is good or not.

It is quite obvious state of the application/client may result in different responses (i.e. more operations invocable by the current user), but having a variable entry point - what purpose could it serve?

Reply:
I don't see a problem here. The entrypoint is just a link after all. It can point to any resource the server chooses to serve as the entrypoint. 

The client should make no assumption about the URL nor about the representation. Hence the server is free to do as it pleases 💪.

> but having a variable entry point - what purpose could it serve?

I would even go further, and add that the server state can also affect the entrypoint (it's URI or representation; doesn't matter which). I just implemented that today in an app that a fresh installation on empty database returns an entrypoint with a `Install` operation which bootstraps some minimum representations. Once executed the "standard" entrypoint is served.

Reply:
That's why I found it interesting :). The example talks to me.